lib/spack/spack/solver/asp.py

```diff
@@ -1700,6 +1770,7 @@ def platform_defaults(self):
         self.gen.h2("Default platform")
         platform = spack.platforms.host()
         self.gen.fact(fn.node_platform_default(platform))
+        self.gen.fact(fn.allowed_platform(platform))
```

Comment:
> I'm guessing this is required or the test_requirements_fail_with_custom_message fails for the test with the platform=test condition?

I'll double check the details, but this was required because we don't treat `platform` like the other node attributes (which have a list of allowed values they can take). So, if we have requirements based on `platform` that cannot be satisfied sometimes Spack would pick a wrong platform depending on unrelated facts in the solve.

Reply:
My assessment is that this is a new hard constraint but that it won't impact users: there are not many platforms and users aren't currently cross compiling other than possibly OS/OS or target/target (in particular since Cray/Linux are now the same platform). My original concern was that a user might be specifying an alternative platform and that this would break that, but I don't think they would get anything out of doing so, and could therefore reconcretize with their local platform without issue.